lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,253 @@
+      # @return [Boolean] success of all csrs being submitted to CA
+      def submit_certificate_requests(certname_csr_hash, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_request')
+
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certname_csr_hash.each do |certname, csr_string|
+            url.resource_name = certname
+            result = connection.put(csr_string, url)
+
+            log_csr_result(result, certname)
+          end
+        end
+      end
+
+      def log_csr_result(result, certname)
+        case result.code
+        when '200', '204'
+          @logger.inform "#{certname} has a waiting certificate request"
+        else
+          @logger.err 'Error:'
+          @logger.err "    When request submitted for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+        end
+      end
+
+      # Make an HTTP request to CA to sign the named certificates
+      # @param certnames [Array] the names of the certificates to have signed
+      # @return [Boolean] the success of all certificates being signed
+      def sign_certs(certnames, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_status')
+
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certnames.each do |certname|
+            url.resource_name = certname
+            body = JSON.dump({desired_state: 'signed'})
+            result = connection.put(body, url)
+            log_sign_result(result, certname)
+          end
+        end
+      end
+
+      def log_sign_result(result, certname)
+        case result.code
+        when '204'
+          @logger.inform "Signed certificate request for #{certname}"
+        else
+          @logger.err 'Error:'
+          @logger.err "    When download requested for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+        end
+      end
+
+      # Make an HTTP request to fetch the named certificates from CA
+      # @param certnames [Array] the names of the certificates to fetch
+      # @return [Boolean] the success of all certificates being dowloaded
+      def download_certificates_from_ca(certnames, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate')
+
+        success = HttpUtilities.with_connection(url, settings) do |connection|
+          certnames.map do |certname|
+            url.resource_name = certname
+            result = connection.get(url)
+            if downloaded = check_download_result(result, certname)
+              save(result.body, certname, settings[:certdir])
```

Comment:
Is it ok to do a save from within a connection?  Could that take too long?  My only other option would be to return a hash of certnames with the cert strings and save them after.